vpr/SRC/route/route_tree_timing.c

```diff
@@ -479,16 +562,600 @@ void free_route_tree(t_rt_node * rt_node) {
+		}
+		// join point (sibling edges)
+		else if (edge->next) {
+			// recursively update for each of its sibling branches
+			t_trace* sink;
+			do {
+				head->iswitch = edge->iswitch;
+				// a sub-branch off the current one, such that branch_head's parent is head
+				t_trace* branch_head = alloc_trace_data();
+				branch_head->index = edge->child->inode;
+				head->next = branch_head;
+
+				// the tip/end of the sub-branch
+				sink = traceback_branch_from_route_tree(branch_head, edge->child, sinks_left);
+				// danger here of losing a pointer if sink came from a branch point since sink->next would already be allocated
+				// don't allocate for last sink (no branches after the last sink)
+				if (sinks_left) {
+					sink->next = alloc_trace_data();	
+					// copy the information to the other instances (1 for each branch start) of the branch point
+					sink->next->index = head->index;
+					head = sink->next;	// become the since each branch might have different switches
+				}
+				edge = edge->next;
+			} while (edge);
+			return sink;	// returns the last ("rightmost") sink rooted at this branch point
+		}
+		// edge must be valid here, and there is only 1 edge
+		head->iswitch = edge->iswitch;
+		root = edge->child;
+		edge = root->u.child_list;
+
+		head->next = alloc_trace_data();
+		// head is now the next head, along with root being the next root
+		head = head->next;
+		head->index = root->inode;
+	}
+}
+t_trace* traceback_from_route_tree(int inet, const t_rt_node* root, int num_routed_sinks) {
+
+	/* Creates the traceback for net inet from the route tree rooted at root *
+	 * properly sets trace_head and trace_tail for this net 				 *
+	 * returns the trace head for inet 					 					 */
+
+	t_trace* head {alloc_trace_data()};
+	head->index = root->inode;
+	// the very last sink
+	auto tail = traceback_branch_from_route_tree(head, root, num_routed_sinks);
+	// should've reached all existing sinks
+	assert(num_routed_sinks == 0);
+
+	// tag end of traceback
+	tail->next = nullptr;
+
+	trace_tail[inet] = tail;
+	trace_head[inet] = head;
+
+	return head;
+}
+
+
+/***************  Pruning and fixing cost of skeleton route tree ****************/
+static void find_sinks_of_route_tree(const t_rt_node* rt_root, vector<int>& sink_nodes) {
+
+	/* Push back the node index of all sinks rooted at rt_root onto sink_nodes */
+
+	t_linked_rt_edge* edge {rt_root->u.child_list};
+	for (;;) {
+		// reached a sink
+		if (!edge) {
+			sink_nodes.push_back(rt_root->inode);
+			return;
+		}
+		// join point (sibling edges)
+		else if (edge->next) {
+			// recursively update for each of its sibling branches
+			do {
+				find_sinks_of_route_tree(edge->child, sink_nodes);
+				edge = edge->next;
+			} while (edge);
+			return;
+		}
+
+		rt_root = edge->child;
+		edge = rt_root->u.child_list;
+	}
+}
+static bool prune_illegal_branches_from_route_tree(t_rt_node* rt_root, float pres_fac, vector<t_rt_node*>& reached_rt_sinks, 
```

Comment:
Have we verified that this produces the same R_upstream / C_downstream as the old code?